process_report/processors/prepayment_processor.py

```diff
@@ -0,0 +1,202 @@
+            row_mask = self.data[invoice.PROJECT_NAME_FIELD].isin(
+                group_dict[invoice.PREPAY_PROJECT_FIELD]
+            )
+            col_mask = [
+                invoice.INVOICE_EMAIL_FIELD,
+                invoice.GROUP_NAME_FIELD,
+                invoice.GROUP_INSTITUTION_FIELD,
+                invoice.GROUP_MANAGED_FIELD,
+            ]
+            self.data.loc[row_mask, col_mask] = [
+                group_dict[invoice.PREPAY_GROUP_CONTACT_FIELD],
+                group_name,
+                group_institute,
+                group_dict[invoice.PREPAY_MANAGED_FIELD],
+            ]
+
+    def _apply_prepayments(self):
+        for group_name, group_dict in self.group_info_dict.items():
+            group_projects = self.data[
+                self.data[invoice.GROUP_NAME_FIELD] == group_name
+            ]
+            prepay_amount_used = self.apply_flat_discount(
+                self.data,
+                group_projects,
+                invoice.PI_BALANCE_FIELD,
+                group_dict[invoice.GROUP_BALANCE_FIELD],
+                invoice.GROUP_BALANCE_USED_FIELD,
+                invoice.BALANCE_FIELD,
+            )
+
+            remaining_prepay_balance = (
+                group_dict[invoice.GROUP_BALANCE_FIELD] - prepay_amount_used
+            )
+            self.data.loc[
+                self.data[invoice.GROUP_NAME_FIELD] == group_name,
+                invoice.GROUP_BALANCE_FIELD,
+            ] = remaining_prepay_balance
+
+            # If the group has used some prepay money, check if the group
+            # already has a debit entry for the current month to decide
+            # whether to append a new debit entry, or overwrite the old one
+            if prepay_amount_used > 0:
+                debit_entry_mask = (
+                    self.prepay_debits[invoice.PREPAY_MONTH_FIELD] == self.invoice_month
+                ) & (self.prepay_debits[invoice.PREPAY_GROUP_NAME_FIELD] == group_name)
+                if self.prepay_debits[debit_entry_mask].empty:
+                    self.prepay_debits = pandas.concat(
+                        [
+                            self.prepay_debits,
+                            pandas.DataFrame(
+                                [[self.invoice_month, group_name, prepay_amount_used]],
+                                columns=self.prepay_debits.columns,
+                            ),
+                        ],
+                        ignore_index=True,
+                    )
+                else:
+                    self.prepay_debits.loc[
+                        debit_entry_mask, invoice.PREPAY_DEBIT_FIELD
+                    ] = prepay_amount_used
+
+    def _export_prepay_debits(self):
+        self.prepay_debits.to_csv(self.prepay_debits_filepath, index=False)
+
+    def _export_s3_prepay_debits(self):
+        invoice_bucket = util.get_invoice_bucket()
+        invoice_bucket.upload_file(self.prepay_debits_filepath, "TODO.csv")  # TODO
```

Comment:
@naved001 @knikolla @joachimweyl I'll amend this after we agree on the S3 filename for the debits file